impeller/entity/contents/content_context.h

```diff
@@ -377,338 +377,338 @@ class ContentContext {
 
   Tessellator& GetTessellator() const;
 
-  std::shared_ptr<Pipeline<PipelineDescriptor>> GetFastGradientPipeline(
+  const Pipeline<PipelineDescriptor>* GetFastGradientPipeline(
```

Comment:
What do you think of wrapping this up in a raw_ptr class.  That way in debug builds we could actually assert we aren't using dangling pointers.

```c++
template<typename T>
class raw_ptr {
public:
  raw_ptr(const std::shared_ptr<T>& ptr) 
     : ptr_(ptr.get())
#if !NDEBUG
      , weak_ptr_(ptr)
#endif
     {}
  T* get() {
#if !NDEBUG
   FML_CHECK(weak_ptr_.lock());
#endif
    return ptr_;
   }
private: 
  T* ptr_;
#if !NDEBUG
  std::weak_ptr<T> weak_ptr_;
#endif
};
```

Reply:
I will never use the class as I never run unopt builds. I don't think we should add infrastructure for the one or two people who do.

Reply:
It is executed by CI though.

Reply:
so are things like ASAN

Reply:
rather than writing our own ASAN code, can we just use ASAN on our tests?

Reply:
If we were running ASAN that sounds good to me.  It looks like the linux_unopt.json is passing in the `--asan` argument to gn and is running `run_tests.py --variant engine`.  However in `run_tests.py` it looks like impeller_unittests are only run on macos.  I don't know if we are getting any ASAN coverage with this code.

Reply:
I think we should be able to.  There is some weirdness when executing on linux.  You have to make a fake X instance.  Looks like we have something in run_tests.py that does this already: https://github.com/gaaclarke/engine/blob/9fff75d9364fb3eafb62240ac1ab04f76ef914cd/testing/run_tests.py#L1220

You have to specify "impeller" as the variant though (we are just specifying engine).  I don't know if it works.  It might not if there are no builders executing it.  I think this goes back all the way to when kaushik hooked up it.

Reply:
We can try adding "impeller" to the variant in linux_unopt.json and see if it works 🤞

Reply:
I can't get things compiling with ASAN on macos:
```
$ ./flutter/tools/gn --unoptimized --runtime-mode=debug --no-rbe --no-goma --asan --mac-cpu=arm64
$ ninja -j8 -C out/host_debug_unopt_arm64 impeller_golden_tests
[5/2329] ACTION //flutter/impeller/runtime_stage:flatc_runtime_stage_flatbuffers(//build/toolchain/mac:clang_arm64)
FAILED: gen/flutter/impeller/runtime_stage/runtime_stage_flatbuffers.h 
vpython3 ../../build/gn_run_binary.py flatc --warnings-as-errors --cpp --cpp-std c++17 --cpp-static-reflection --gen-object-api --filename-suffix _flatbuffers -o gen/flutter/impeller/runtime_stage ../../flutter/impeller/runtime_stage/runtime_stage.fbs
Command failed: ./flatc --warnings-as-errors --cpp --cpp-std c++17 --cpp-static-reflection --gen-object-api --filename-suffix _flatbuffers -o gen/flutter/impeller/runtime_stage ../../flutter/impeller/runtime_stage/runtime_stage.fbs
exitCode: -6
dyld[61885]: missing symbol called
```

Reply:
I filed an issue for ASAN: https://github.com/flutter/flutter/issues/159985